### ISSUE TITLE
[Bug] [DFlash2] Greedy output diverges from target-only Qwen3.8-27B when thinking is enabled

### ISSUE THREAD

**Issue by forlayo** [sgl-project/sglang#38009]
### Checklist

- [x] I searched related issues but found no solution.
- [x] The bug persists in the latest version.
- [x] Issues without environment info and a minimal reproducible demo are hard to resolve and may receive no feedback.
- [x] If this is not a bug report but a general question, please start a discussion at https://github.com/sgl-project/sglang/discussions. Otherwise, it will be closed.
- [x] Please use English. Otherwise, it will be closed.

## Summary

Using SGLang's dedicated Qwen3.8/DFlash2 development image, greedy decoding
with DFlash2 does not produce the same token sequence as the target-only model
when Qwen3.8 thinking is enabled. The first divergence is deterministic at
completion token 30 for the reproduction below.

The target-only baseline is deterministic across two identical requests, and
the DFlash2 result is exactly equal to the target-only result with thinking
disabled. This makes the issue appear specific to the DFlash/hybrid
state/thinking path rather than prompt formatting, sampling, or general model
loading.

## Expected behavior

For greedy decoding (`temperature=0`), DFlash2 speculative decoding should
produce exactly the same generated token sequence as target-only decoding for
the same model, prompt, chat-template settings, and seed.

## Observed behavior

With thinking enabled, the sequences diverge at generated token index 30:

| Profile | Tokens around the first difference |
| --- | --- |
| target-only | `with canvas, particles, cat-themed pieces (cat` |
| DFlash2 | `with canvas, particles, Web Audio API sounds,` |

The corresponding `reasoning_content` continues differently after that point.

The exact comparison summary is:

```json
{
  "baseline_repeat": {"exact": true, "first_difference": null},
  "thinking": {
    "exact": false,
    "first_difference": 30,
    "baseline_token": " cat",
    "dflash_token": " Web"
  },
  "no_thinking": {"exact": true, "first_difference": null}
}
```

## Environment

- Host: native Linux, 4 × NVIDIA RTX 3090 (24 GiB each), tensor parallelism 4.
- SGLang image: `lmsysorg/sglang:dev-cu13-qwen38-27b-dflash2`
- Image digest: `sha256:616a3e97f45191af975896cfa644279096cb31bd408a071c2e99ca7209c3cafe`
- SGLang commit from image label: `5f55db35e926d50676f75b812640ea2410b0fe0e`
- SGLang version: `0.0.0.dev1+g5f55db35e`
- Container CUDA: 13.0.3; NCCL reported by SGLang: 2.30.7.
- Target: local copy of `Qwen/Qwen3.8-27B`, BF16.
- Draft: local copy of `incoai/Qwen3.8-27B-DFlash2`, BF16.
- Tensor parallel: `--tp-size=4`.
- DFlash block size: 8.
- KV dtype: SGLang default (`auto`, BF16 in this BF16 run); no KV quantization.

The validation intentionally used a single request, 32k context and a reduced
static memory fraction only to fit the target and draft on 24-GiB cards:

```text
--context-length=32768
--mem-fraction-static=0.75
--max-running-requests=1
--disable-radix-cache
--disable-overlap-schedule
--disable-cuda-graph
```

Those flags are identical in baseline and DFlash profiles. They remove cache,
scheduling and graph-capture variables; they do not alter sampling or logits.

## Server commands

Baseline profile:

```bash
python3 -m sglang.launch_server \
  --model-path /root/.cache/huggingface/models/Qwen3.8-27B \
  --served-model-name qwen3.8-27b \
  --tp-size 4 --host 0.0.0.0 --port 30000 \
  --context-length 32768 --mem-fraction-static 0.75 \
  --max-running-requests 1 \
  --reasoning-parser qwen3 --tool-call-parser qwen3_coder \
  --disable-radix-cache --disable-overlap-schedule --disable-cuda-graph
```

DFlash profile differs only by adding:

```bash
--speculative-algorithm DFLASH \
--speculative-draft-model-path /root/.cache/huggingface/models/Qwen3.8-27B-DFlash2 \
--speculative-dflash-block-size 8
```

## Reproduction request

Send this request to each profile (target-only twice, then DFlash once):

```json
{
  "model": "qwen3.8-27b",
  "messages": [{
    "role": "user",
    "content": "hazme un tetris de gatos en una pagina web, debe de ser muy impactante y visualmente atractivo con efecto gráficos y sonoros"
  }],
  "temperature": 0,
  "top_p": 1,
  "top_k": -1,
  "seed": 20260902,
  "max_tokens": 64,
  "logprobs": true,
  "top_logprobs": 5,
  "chat_template_kwargs": {
    "reasoning_effort": "medium",
    "preserve_thinking": true
  }
}
```

For the control, replace `chat_template_kwargs` with:

```json
{"enable_thinking": false}
```

Compare `choices[0].logprobs.content[*].token`, rather than rendered text.
The first 30 tokens match in the thinking run; the next token differs exactly
as shown above. With thinking disabled, all 64 returned tokens match.

## Attached evidence

Attach the six files from
`resultados/20260904T105622Z-qwen38-sglang-dflash2-ab/`:

- `baseline_thinking_a.json`
- `baseline_thinking_b.json`
- `baseline_no_thinking.json`
- `dflash_thinking.json`
- `dflash_no_thinking.json`
- `summary.json`

The requests are local/direct to SGLang; LiteLLM was not in the request path.

## Question

Is this a known limitation or bug in the Qwen3.8 DFlash2 integration when the
model's reasoning/thinking path is active? If a different DFlash setting,
draft checkpoint revision, attention backend, or Mamba/GDN cache configuration
is required for greedy equivalence, please point to the supported configuration.

[dflash_thinking.json](https://github.com/user-attachments/files/31834739/dflash_thinking.json)
[baseline_thinking_b.json](https://github.com/user-attachments/files/31834743/baseline_thinking_b.json)
[baseline_thinking_a.json](https://github.com/user-attachments/files/31834741/baseline_thinking_a.json)
[baseline_no_thinking.json](https://github.com/user-attachments/files/31834740/baseline_no_thinking.json)
[summary.json](https://github.com/user-attachments/files/31834742/summary.json)
[dflash_no_thinking.json](https://github.com/user-attachments/files/31834744/dflash_no_thinking.json)

### Reproduction

Send this request to each profile (target-only twice, then DFlash once):

```json
{
  "model": "qwen3.8-27b",
  "messages": [{
    "role": "user",
    "content": "hazme un tetris de gatos en una pagina web, debe de ser muy impactante y visualmente atractivo con efecto gráficos y sonoros"
  }],
  "temperature": 0,
  "top_p": 1,
  "top_k": -1,
  "seed": 20260902,
  "max_tokens": 64,
  "logprobs": true,
  "top_logprobs": 5,
  "chat_template_kwargs": {
    "reasoning_effort": "medium",
    "preserve_thinking": true
  }
}
```

For the control, replace `chat_template_kwargs` with:

```json
{"enable_thinking": false}
```

Compare `choices[0].logprobs.content[*].token`, rather than rendered text.
The first 30 tokens match in the thinking run; the next token differs exactly
as shown above. With thinking disabled, all 64 returned tokens match.

### Environment

- Host: native Linux, 4 × NVIDIA RTX 3090 (24 GiB each), tensor parallelism 4.
- SGLang image: `lmsysorg/sglang:dev-cu13-qwen38-27b-dflash2`
- Image digest: `sha256:616a3e97f45191af975896cfa644279096cb31bd408a071c2e99ca7209c3cafe`
- SGLang commit from image label: `5f55db35e926d50676f75b812640ea2410b0fe0e`
- SGLang version: `0.0.0.dev1+g5f55db35e`
- Container CUDA: 13.0.3; NCCL reported by SGLang: 2.30.7.
- Target: local copy of `Qwen/Qwen3.8-27B`, BF16.
- Draft: local copy of `incoai/Qwen3.8-27B-DFlash2`, BF16.
- Tensor parallel: `--tp-size=4`.
- DFlash block size: 8.
- KV dtype: SGLang default (`auto`, BF16 in this BF16 run); no KV quantization.

The validation intentionally used a single request, 32k context and a reduced
static memory fraction only to fit the target and draft on 24-GiB cards:

```text
--context-length=32768
--mem-fraction-static=0.75
--max-running-requests=1
--disable-radix-cache
--disable-overlap-schedule
--disable-cuda-graph
```

Those flags are identical in baseline and DFlash profiles. They remove cache,
scheduling and graph-capture variables; they do not alter sampling or logits.

**Comment by yangqinning** [comment]
### Same engine image (digest `616a3e97…`) on DGX Spark: DFlash2 greedy output in thinking mode is reproducible within one process, but not across server restarts

Not a reproduction of the exact token-30 divergence (different hardware, different target quantization), but directly relevant to how the comparison is drawn: on our box the DFlash2 thinking output itself is not stable across process restarts, so a target-only run from one server start and a DFlash2 run from another are not guaranteed to be comparable.

**Environment** (same engine image as reported here):

- `lmsysorg/sglang:qwen38-27b-dflash2`, digest `sha256:616a3e97f45191af975896cfa644279096cb31bd408a071c2e99ca7209c3cafe`, SGLang `0.0.0.dev1+g5f55db35e`
- NVIDIA DGX Spark (GB10, single GPU, aarch64/SM121, 121 GiB unified memory)
- Target `RadixArk/Qwen3.8-27B-NVFP4-BF16-LMHead` (NVFP4 W4A4, dense BF16 `lm_head`), draft `z-lab/Qwen3.8-27B-DFlash2`, DFLASH block 8, FP8 KV cache
- One request at a time, `temperature=0`, thinking on (default), ~900 generated tokens, streamed; prompt is a Chinese code-generation request

**Observed:**

1. Within one server process, repeated identical requests return byte-identical text (deterministic).
2. Across container restarts with identical flags, the same request diverges: the first differing character is near the start of the streamed thinking text, at byte offsets ~128 (a block-8 pair) and 67 / 80 (separate block-16 boots). No request-level repetition produced a difference; the difference only appears between processes.
3. The divergence correlates with a change in accepted length: most boots accepted 2.5–3.5 tokens per step (20.1–21.6 tok/s), while one block-16 boot accepted 4.4–6.2 tokens per step (30.1 tok/s).
4. Block size changes the output as well (8 vs 16 differ), but the cross-boot instability is present at the default block 8 too.

**Not tested:** whether a target-only (non-speculative) process is also unstable across restarts in this build. If it is stable, the instability here is specific to the DFlash2 path and the original comparison stands as reported. If it also drifts, then comparing target-only from one server start against DFlash2 from another cannot attribute the divergence to DFLASH alone — a same-process comparison, or A/A runs across restarts for each profile, would be needed.

**Suggestion for the repro:** run each profile in two fresh server starts and compare A/A per profile before the A/B (target-only vs DFlash2). Logging `meta_info`/accept length per request is also worth capturing: the per-boot accept-length swings here (2.5–3.5 → 4.4–6.2) suggest boots can settle into different numerical behavior, which would surface as exactly this kind of token difference.
